### PR TITLE
simplifying callout editor in levelbuilder

### PR DESCRIPTION
**Pull request by tanyaparker**

### CODE REVIEW

File: dashboard/app/views/levels/editors/_blockly.html.haml [hunk 1]
```diff
@@ -296,269 +171,98 @@
...
-    updated_callouts_2.qtip_config.position.at = $('.qtip_position_at_2').val();
-  if ($('.qtip_position_adjust_x_2').val().length > 0)
-    updated_callouts_2.qtip_config.position.adjust.x = $('.qtip_position_adjust_x_2').val();
-  if ($('.qtip_position_adjust_y_2').val().length > 0)
-    updated_callouts_2.qtip_config.position.adjust.y = $('.qtip_position_adjust_y_2').val();
-  if ($(".qtip_hide_2").prop("checked") == true)
-    updated_callouts_2.qtip_config.style.classes = "no-tip"
+  var updated_callouts = [updated_callouts_0, updated_callouts_1, updated_callouts_2];
+
+  for (var i=0; i<3; i++) {
```
Comment (bcjordan) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: `updated_callouts.length` in place of `3` here

**Comment by joshlory** [comment]
Only need 1 change handler, could bind to e.g. `'.callouteditor input'`

**Comment by bcjordan** [comment]
Yeah, using the update handler Josh said, it should be fine performance-wise to just always update the hash representation on every input change.

& if you want to loop over references to those local variables like `updated_callouts_0`, you could have those operations work on the `updated_callouts[n]` full array instead

File: dashboard/app/views/levels/editors/_blockly.html.haml [hunk 2]
```diff
@@ -78,193 +70,76 @@
...
-        %td
-          %input{:class => "qtip_position_adjust_x_2", :type => 'number'}
-        %td
-          %input{:class => "qtip_position_adjust_y_2", :type => 'number'}
-        %td
-          %input{:class => "qtip_hide_2", :type => 'checkbox'}
+  - (0...3).each do |x|
+    %label Callout #{x+1}
+
+    %div{:class => "callouts#{x}"}
```
Comment (joshlory) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: Nit prefer `{class: "callouts#{x}"}` over `=>` for symbol keys.

File: dashboard/app/views/levels/editors/_blockly.html.haml [hunk 3]
```diff
@@ -296,269 +171,98 @@
...
-        $(".qtip_position_at_2").val(parsecallouts[2].qtip_config.position.at)
-      if ((parsecallouts[2].qtip_config) && (parsecallouts[2].qtip_config.position) && (parsecallouts[2].qtip_config.position.adjust))
-        $(".qtip_position_adjust_x_2").val(parsecallouts[2].qtip_config.position.adjust.x)
-      if ((parsecallouts[2].qtip_config) && (parsecallouts[2].qtip_config.position) && (parsecallouts[2].qtip_config.position.adjust))
-        $(".qtip_position_adjust_y_2").val(parsecallouts[2].qtip_config.position.adjust.y)
-      if ((parsecallouts[2].qtip_config) && (parsecallouts[2].qtip_config.style) && (parsecallouts[2].qtip_config.style.classes) && (parsecallouts[2].qtip_config.style.classes == "no-tip"))
-        $(".qtip_hide_2").prop("checked", true)
+    for (var i=0; i<callouts.length; i++) {
+      if (parsecallouts[i]) {
+        $(".use_callout_"+i).prop("checked", true)
```
Comment (joshlory) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: Nit: spacing around `+`, e.g. `$(".use_callout_" + i)`

File: dashboard/app/views/levels/editors/_blockly.html.haml (hunk 2 above)
Comment (joshlory) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: Note this doesn't block check in at all, and if you prefer I can fix this up from RubyMine (1 click to update all occurrences in the file).

Comment (tanyaparker) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: No worries! Quick fix. Anything else?

**Comment by joshlory** [comment]
Rather than having 3 fixed callout slots, can there just be 1 to start and an 'Add Callout' button to add a new row?

**Comment by joshlory** [comment]
If we go that route the "Use callout?" field can go away.

File: dashboard/app/views/levels/editors/_blockly.html.haml [hunk 4]
```diff
@@ -78,193 +70,76 @@
...
+              %option{value: 'left top'} Left top  
+          %td
+            %input{class: "qtip_position_adjust_x_#{x}", type: 'number'}
+          %td
+            %input{class: "qtip_position_adjust_y_#{x}", type: 'number'}
+    - unless x==2
+      %hr
+    
 
 .field
```
Comment (joshlory) [inline dashboard/app/views/levels/editors/_blockly.html.haml]: This whole field can go away now that the new editor is in place.

**Comment by tanyaparker** [comment]
Yeah, you're right. It should be dynamic. Might as well fix that now too.

**Comment by joshlory** [comment]
Yeah sorry, I realize this is starting from existing code.  I think dynamic will make the JS cleaner as well.

**Comment by tanyaparker** [comment]
Going to merge this clean-up so it's easier to read the file. TODO: make dynamic instead of only 3 callouts.

**Pull request merged**